replicator/active_replicator_pull.go

```diff
@@ -69,19 +162,163 @@ func (apr *ActivePullReplicator) Start() error {
+	}
+
+	// checkpointerLock guards the expectedSeqs slice, and receivedSeqs map
+	checkpointerLock := sync.Mutex{}
+
+	// An ordered list of sequence IDs we expect to receive revs for.
+	expectedSeqs := []db.SequenceID{}
+	// A map of sequence IDs we've received revs for.
+	receivedSeqs := map[db.SequenceID]struct{}{}
+
+	checkpointNow := make(chan struct{})
+
+	apr.blipSyncContext.postHandleChangesCallback = func(changesSeqs []db.SequenceID) {
+		apr.Stats.Add(ActiveReplicatorStatsKeyChangesRevsReceivedTotal, int64(len(changesSeqs)))
+
+		if len(changesSeqs) == 0 {
+			// nothing to do
+			return
+		}
+
+		select {
+		case <-apr.ctx.Done():
+			// replicator already closed, bail out of checkpointing work
+			return
+		default:
+		}
+
+		checkpointerLock.Lock()
+		expectedSeqs = append(expectedSeqs, changesSeqs...)
+		checkpointerLock.Unlock()
+	}
+
+	revCounter := uint16(0)
+	apr.blipSyncContext.postHandleRevCallback = func(remoteSeq db.SequenceID) {
+		apr.Stats.Add(ActiveReplicatorStatsKeyRevsReceivedTotal, 1)
+
+		select {
+		case <-apr.ctx.Done():
+			// replicator already closed, bail out of checkpointing work
+			return
+		default:
+		}
+
+		checkpointerLock.Lock()
+		revCounter++
+		receivedSeqs[remoteSeq] = struct{}{}
+
+		shouldCheckpoint := revCounter >= apr.config.CheckpointRevCount
+		if !shouldCheckpoint {
+			checkpointerLock.Unlock()
+			return
+		}
+
+		revCounter = uint16(0)
+		checkpointerLock.Unlock()
+
+		select {
+		case checkpointNow <- struct{}{}:
+		case <-time.After(apr.config.CheckpointMinInterval):
+			// Blocked trying to send a checkpointNow signal
+			base.Tracef(base.KeyReplicate, "checkpointer: skipping sending blocked checkpointNow signal")
+		}
+	}
+
+	// Start a time-based checkpointer goroutine
+	go func() {
+		timeout := time.NewTimer(apr.config.CheckpointMaxInterval)
+		defer timeout.Stop()
+		for {
+			select {
+			case <-checkpointNow:
+				base.Tracef(base.KeyReplicate, "checkpointer: got checkpointNow signal")
+				// fall out of select to set the checkpoint
+			case <-timeout.C:
+				base.Tracef(base.KeyReplicate, "checkpointer: got timeout signal")
+				// timed out waiting for number of revs to reach CheckpointRevCount
+				// fall out of select to set the checkpoint
+			case <-apr.ctx.Done():
+				// replicator stopped, set a final checkpoint
+			}
+
+			apr.checkpointerWg.Add(1)
+			if !timeout.Stop() {
+				<-timeout.C
+			}
+
+			checkpointerLock.Lock()
+
+			// find the highest contiguous sequence we've received
+			if len(expectedSeqs) > 0 {
+				var lowSeq db.SequenceID
+
+				// iterates over each (ordered) expected sequence and stops when we find the first sequence we've yet to receive a rev message for
+				var maxI int
+				for i, seq := range expectedSeqs {
+					if _, ok := receivedSeqs[seq]; !ok {
+						base.Tracef(base.KeyReplicate, "checkpointer: couldn't find %v in receivedSeqs", seq)
+						break
+					}
+
+					delete(receivedSeqs, seq)
+					maxI = i
+				}
+
```

Comment:
This looks like it's setting lowSeq to expectedSeqs[0] if we don't find expectedSeqs[0] in receivedSeqs.